docs/USAGE-GUIDE.md

```diff
@@ -1112,30 +1112,58 @@ const char * s2n_connection_get_curve(struct s2n_connection *conn);
 
 **s2n_connection_get_curve** returns a string indicating the elliptic curve used during ECDHE key exchange. The string "NONE" is returned if no curve has was used.
 
-### Session State Related calls
+### Session Resumption Related calls
 
 ```c
+int s2n_config_set_session_state_lifetime(struct s2n_config *config, uint32_t lifetime_in_secs);
+
 int s2n_connection_set_session(struct s2n_connection *conn, const uint8_t *session, size_t length);
 int s2n_connection_get_session(struct s2n_connection *conn, uint8_t *session, size_t max_length);
 ssize_t s2n_connection_get_session_length(struct s2n_connection *conn);
 ssize_t s2n_connection_get_session_id_length(struct s2n_connection *conn);
 int s2n_connection_is_session_resumed(struct s2n_connection *conn);
 ```
 
-- **session** session will contain serialized session related information needed to resume handshake.
+- **lifetime_in_secs** lifetime of the cached session state required to resume a handshake
+- **session** session will contain serialized session related information needed to resume handshake either using session id or session ticket.
 - **length** length of the serialized session state.
 - **max_length** Max number of bytes to copy into the **session** buffer.
 
-**s2n_connection_set_session** de-serializes the session state and updates the connection accrodingly.
+**s2n_config_set_session_state_lifetime** sets the lifetime of the cached session state. The default value is 15 hours.
+
+**s2n_connection_set_session** de-serializes the session state and updates the connection accordingly.
 
-**s2n_connection_get_session** serializes the session state from connection and copies into the **session** buffer and returns the number of bytes that were copied.
+**s2n_connection_get_session** serializes the session state from connection and copies into the **session** buffer and returns the number of bytes that were copied. If the first byte in **session** is 1, then the next 4 bytes will contain the session ticket lifetime hint in seconds.
 
-**s2n_connection_get_session_length** returns number of bytes needed to store serailized session state; it can be used to allocate the **session** buffer.
+**s2n_connection_get_session_length** returns number of bytes needed to store serialized session state; it can be used to allocate the **session** buffer.
 
 **s2n_connection_get_session_id_length** returns session id length from the connection.
 
 **s2n_connection_is_session_resumed** checks if the handshake is abbreviated or not.
 
+### Session Ticket Specific calls
+
+```c
+int s2n_config_disable_session_tickets(struct s2n_config *config);
+int s2n_config_set_ticket_valid_key_lifetime(struct s2n_config *config, uint64_t lifetime_in_secs);
+int s2n_config_set_ticket_semi_valid_key_lifetime(struct s2n_config *config, uint64_t lifetime_in_secs);
+int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *name, uint32_t name_len, uint8_t *key, uint32_t key_len);
```

Comment:
How can we support server restarts with s2n API? In particular: i have my server running with some keys in place, some of which are semi-valid, and 1 valid one. How can I restart my server without loosing semi-valid keys for decryption? Right now that doesn't seem to be possible, so I wonder if we should make a slightly different API to pass valid and semi-valid lifetime together with key.

Reply:
Each key having it's own valid lifetime would mess up the probability logic of selecting a valid key from multiple valid keys. The duration of valid lifetime has to be same for all keys when computing probability for https://github.com/awslabs/s2n/issues/242 to work.